### DISCUSSION TITLE
Native Rust Implementation

### DISCUSSION THREAD

**Discussion by solaeus** [tree-sitter/tree-sitter#2848] [Ideas] [answered]
Do the maintainers have any thoughts on porting the project to Rust? I've been working with Tree Sitter for about six months and have run into strange infinite loops when generating parsers and the issues show a history of C-related programming problems. A rust implementation would improve the correctness of the program and allow tree-sitter to play nicely with serde and the Rust standard library (I'm especially disappointed at the lack of multi-thread support).

**Comment by maxbrunsfeld** [comment, marked as answer]
Multi threading is fully supported.

**Reply by solaeus** [reply]
I have no issues sharing Tree instances between threads but I can't share Node instances.

**Reply by maxbrunsfeld** [reply]
Nodes are references to parts of trees, so it’s not intended to share these between threads.

this decision is orthogonal to what language we use to implement the library in though.

**Reply by solaeus** [reply]
So why is it conceptually incorrect to share parts of trees between threads?

**Comment by code-kungfu** [comment]
That would not make any sense at all.
There's already too much Rust in there.

It isn't a failure of the C programming language but rather you're using trees-sitter wrong.
It works perfectly in a multi-threaded setting.

**Reply by solaeus** [reply]
EDIT: Judging from the content of this comment and the commenter's profile, this is a troll.
